Memgraph in high throughput workloads (#1266)
memgraph/documentation: pages/memgraph-in-production/memgraph-in-high-throughput-workloads.mdx

## Is this guide for you?

This guide is for you if you're working with **high-throughput graph workloads** where performance, consistency,
and scale are critical.
You’ll benefit from this content if:

- ⚡ You’re handling **more than a thousand writes per second**, and your graph data is constantly changing at high velocity.  
- 🔍 You want your **read performance to remain consistent**, even as new data is continuously ingested.  
- 🔁 You’re dealing with **high volumes of concurrent reads and writes**, and need a database that can handle both without performance degradation.  
- 🌊 Your data is flowing in from **real-time streaming systems** like **Kafka**, and you need a database that can keep up.  

If this sounds like your use case, this guide will walk you through how to configure and scale Memgraph for **reliable, high-throughput performance** in production.

## Why choose Memgraph for high-throughput use cases?

When your workload involves thousands of writes per second and concurrent access to ever-changing graph data, 
Memgraph provides the performance and architecture needed to keep up—without compromise. 

Here's why Memgraph is a great fit for high-throughput use cases:

> - **In-memory storage engine**  
>   Memgraph doesn't need to write to disk on every transaction, enabling it to **scale write throughput far beyond traditional 
>   disk-based databases**. Unlike systems that rely on LRU or OS-level caching—where **cache invalidation can degrade
>   read performance during heavy writes**, Memgraph offers **predictable read latency** even under constant data changes.  
>   
>   While many graph databases **max out around 1,000 writes per second**, Memgraph can handle **up to 50x more** (image below),
>   making it ideal for **high-velocity, write-intensive workloads**. 
> 
>   ![](/pages/memgraph-in-production/benchmarking-memgraph/realistic-workload.png)
> 
> - **Non-blocking reads and writes with MVCC**  
>   Built on **multi-version concurrency control (MVCC)**, Memgraph ensures that **writes don’t block reads** and
>   **reads don’t block writes**, allowing each to scale independently.
> 
> - **Fine-grained locking**  
>   Locking happens at the **node and relationship level**, enabling **highly concurrent writes** and minimizing
>   contention across threads.
> 
> - **Lock-free skiplist storage**  
>   Memgraph uses **lock-free, concurrent skiplist structures** for storing nodes, relationships, and indices, leading to
>   faster data access and minimal coordination overhead between threads.
> 
> - **Snapshot isolation by default**  
>   Unlike many databases that rely on **read-committed** isolation (which can return inconsistent data), Memgraph provides
>   **snapshot isolation**, ensuring data accuracy and consistency in real-time queries.
> 
> - **Inter-query parallelization**  
>   Each read and write query is handled on its own CPU core, meaning Memgraph can **scale horizontally on a
>   single machine** based on your hardware.
> 
> - **Horizontal read scaling with high availability**  
>   Memgraph supports **replication and high availability**, allowing you to distribute **read traffic across multiple replicas**.
>   These replicas can also power **secondary workloads** like GraphRAG, analytics, or ML pipelines, **without affecting

Reviewer: ```suggestion
- **In-memory storage engine**: Memgraph operates entirely in-memory, eliminating the need to write to disk on every transaction. This allows it to **scale write throughput far beyond traditional 
  disk-based databases**. Unlike systems that rely on LRU or OS-level caching, where **cache invalidation can degrade
  read performance during heavy writes**, Memgraph offers **predictable read latency** even under constant data changes.  
  
  While many graph databases **max out around 1,000 writes per second**, Memgraph can handle **up to 50x more** (see image below),
  making it ideal for **high-velocity, write-intensive workloads**. 

  ![](/pages/memgraph-in-production/benchmarking-memgraph/realistic-workload.png)

- **Non-blocking reads and writes with MVCC**: Built on multi-version concurrency control (MVCC), Memgraph ensures that **writes don’t block reads** and
  **reads don’t block writes**, allowing each to scale independently.

- **Fine-grained locking** : Locking happens at the node and relationship level, enabling **highly concurrent writes** and minimizing contention across threads.

- **Lock-free skiplist storage**: Memgraph uses **lock-free, concurrent skip list structures** for storing nodes, relationships, and indices, leading to
  faster data access and minimal coordination overhead between threads.

- **Snapshot isolation by default**: Unlike many databases that rely on **read-committed** isolation (which can return inconsistent data), Memgraph provides
  **snapshot isolation**, ensuring data accuracy and consistency in real-time queries.

- **Inter-query parallelization**: Each read and write query is handled on its own CPU core, meaning Memgraph can **scale horizontally on a single machine** based on your hardware.

- **Horizontal read scaling with high availability**: Memgraph supports [replication](/clustering/replication) and [high availability](/clustering/high-availability), allowing you to distribute **read traffic across multiple replicas**.
  These replicas can also power **secondary workloads** like GraphRAG, analytics, or ML pipelines, **without affecting

```

streaming use cases.

## Choosing the right Memgraph storage mode

High-throughput scenarios in Memgraph can run effectively on both `IN_MEMORY_TRANSACTIONAL` and `IN_MEMORY_ANALYTICAL`
storage modes, depending on your specific workload requirements.

If your workload meets the following conditions:

- You are **updating properties** on existing nodes and relationships,  
- You are **appending** new nodes and relationships to the graph,  
- You are **not performing deletes**,

then it may be worth considering switching to `IN_MEMORY_ANALYTICAL` mode.  
This mode allows **writes to be multithreaded**, unlocking **near limitless write speeds** by parallelizing ingestion
across CPU cores.

However, keep in mind:

- If you require **replication**, **high availability**, or **ACID guarantees**, you must use `IN_MEMORY_TRANSACTIONAL` mode.  
- `IN_MEMORY_ANALYTICAL` is optimized for **bulk ingestion** and **read-only analytics**, but it 
  **does not support transactional rollback**, as it doesn’t create delta objects during writes.  
  Additionally, **WALs (write-ahead logs) are not generated** in this mode, meaning recovery relies solely on **snapshot creation**.

> 

Reviewer: ```suggestion

Learn more about [storage modes](/fundamentals/storage-memory-usage#storage-modes) in our documentation. 

```

## Importing mechanisms

> Memgraph natively supports **Apache Kafka** and **Apache Pulsar** consumers to directly ingest data streams
> ([learn more here](/data-streams#create-a-stream)). However, depending on the system architecture and flexibility

Reviewer: ```suggestion
Memgraph natively supports **Apache Kafka** and **Apache Pulsar** consumers to directly ingest [data streams](/data-streams). However, depending on the system architecture and flexibility
```

  ensuring your graph remains manageable and efficient over time.

- **Multi-tenancy**  
  Some workloads require a **separate graph database per customer** to ensure strict data isolation and security.
  Memgraph supports [multi-tenancy](/database-management/multi-tenancy), enabling you to manage multiple independent
  graphs within a single Memgraph instance.

## Queries that best suit your workload

When ingesting data from streams, it's best to keep your Cypher queries **simple, idempotent, and efficient**. A typical ingestion query should look like:

```cypher
MERGE (n:Label) SET n += $row;
```

This approach ensures **idempotency** (safe reprocessing of the same events) and **minimizes query execution time** by keeping the transaction lightweight.  
Keep in mind that adding **complex business logic** or **customization** to the ingestion queries will **increase query latency**, so it's always a good practice to **profile your queries** using Memgraph’s [`PROFILE` tool](/querying/clauses/profile) to understand and optimize performance.

---

### Dynamic labels and edge types

Memgraph also supports:

> - [**Dynamic node label creation**](/querying/clauses/create#14-creating-node-labels-dynamically)  
> - [**Dynamic relationship type creation**](/querying/clauses/create#23-creating-relationship-types-dynamically)

Reviewer: ```suggestion
- [Dynamic node label creation](/querying/clauses/create#14-creating-node-labels-dynamically)  
- [Dynamic relationship type creation](/querying/clauses/create#23-creating-relationship-types-dynamically)

```

However, **dynamic creation is only supported with `CREATE` operations**, and **matching or merging dynamically created
labels and types is not supported**.

If your payload contains **dynamic labels or edge types** and you still need **idempotency**, you have two options:

- **Programmatically construct your Cypher query strings** based on the payload to ensure correct label/type usage before sending the query to Memgraph.

> - **Optionally use the `merge` procedure from MAGE** ([MAGE Merge documentation](/advanced-algorithms/available-algorithms/merge))

Reviewer: ```suggestion
- **Optionally use the [`merge`](/advanced-algorithms/available-algorithms/merge) procedure from MAGE**

```
